### ISSUE TITLE
why I get low accuracy in tensorflow serving 

### ISSUE THREAD

**Issue by Aurora11111** [tensorflow/serving#986]
when I run the text classification model in local ,I get accuracy around 95%,but after I export it,and run in tensorflow serving ,I get accuracy around 0.05%,it frustrate me a lot! my code of export server and ,client here:

1. export server 
import tensorflow as tf
import os

SAVE_PATH = './trained_model_1531212474/checkpoints'
MODEL_NAME = 'model'
VERSION = 1
SERVE_PATH = './{}/{}'.format(MODEL_NAME, VERSION)

checkpoint = tf.train.latest_checkpoint(SAVE_PATH)

tf.reset_default_graph()

with tf.Session() as sess:
    # import the saved graph
    saver = tf.train.import_meta_graph(checkpoint + '.meta')
    # get the graph for this session
    graph = tf.get_default_graph()
    sess.run(tf.global_variables_initializer())
    # get the tensors that we need
    print([n.name for n in tf.get_default_graph().as_graph_def().node])
    builder = tf.saved_model.builder.SavedModelBuilder(SERVE_PATH)

    model_inputx = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('input_x:0'))
    model_inputy = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('input_y:0'))
    model_dropout = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('dropout_keep_prob:0'))
    model_accuracy = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('accuracy/accuracy:0'))
    model_predictions = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('output/predictions:0'))

    prediction_signature = tf.saved_model.signature_def_utils.build_signature_def(
        inputs={'input_x': model_inputx,
                'input_y':model_inputy,

                'dropout_keep_prob':model_dropout
                },
        outputs={'output/predictions': model_predictions,
                 'accuracy/accuracy': model_accuracy
                 },
        method_name= tf.saved_model.signature_constants.PREDICT_METHOD_NAME)

    # tensor_info_x = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('input_x:0'))
    # tensor_info_y = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('output/predictions:0'))

    # prediction_signature = (
    #     tf.saved_model.signature_def_utils.build_signature_def(
    #         inputs={'input_x': tensor_info_x},
    #         outputs={'output/predictions': tensor_info_y},
    #         method_name=tf.saved_model.signature_constants.PREDICT_METHOD_NAME))
    legacy_init_op = tf.group(tf.tables_initializer(),
                              name='legacy_init_op')
    builder.add_meta_graph_and_variables(
        sess,
        [tf.saved_model.tag_constants.SERVING],
        signature_def_map={'predict_label': prediction_signature},
        legacy_init_op=legacy_init_op
    )
    # Save the model so we can serve it with a model server :)
    builder.save()

2. client

from grpc.beta import implementations
from tensorflow_serving.apis import predict_pb2
from tensorflow_serving.apis import prediction_service_pb2
import time
import os
import json
import logging
import data_helper
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow.contrib import learn
from data_helper import clean_str

tf.app.flags.DEFINE_string('server', 'localhost:9000',
                           'PredictionService host:port')

FLAGS = tf.app.flags.FLAGS

def numpy_array_int32(data):
  (im_width, im_height) = data.shape
  return np.array(data).reshape(
      (im_width,im_height)).astype(np.int32)

def numpy_array_float(data):
  (im_width, im_height) = data.shape
  return np.array(data).reshape(
      (im_width,im_height)).astype(np.float32)

def main(_):
  print(time.ctime())
  host, port = FLAGS.server.split(':')
  channel = implementations.insecure_channel(host, int(port))
  stub = prediction_service_pb2.beta_create_PredictionService_stub(channel)

  #Send request
  params = json.loads(open('./parameters.json').read())
  checkpoint_dir = './trained_model_1531212474/'
  if not checkpoint_dir.endswith('/'):
      checkpoint_dir += '/'

  """Step 1: load data for prediction"""
  test_file = './data/data.csv'
  test_examples = pd.read_csv(test_file, dtype={'data': object})
  df = pd.read_csv(test_file, dtype={'data': object})
  selected = ['data', 'tag']
  non_selected = list(set(df.columns) - set(selected))

  df = df.drop(non_selected, axis=1)  # Drop non selected columns
  df = df.dropna(axis=0, how='any', subset=selected)  # Drop null rows
  df = df.reindex(np.random.permutation(df.index))  # Shuffle the dataframe
 
  labels = sorted(list(set(df[selected[1]].tolist())))
  one_hot = np.zeros((len(labels), len(labels)), int)
  np.fill_diagonal(one_hot, 1)
  label_dict = dict(zip(labels, one_hot))

  vocab_path = os.path.join(checkpoint_dir, "vocab.pickle")
  vocab_processor = learn.preprocessing.VocabularyProcessor.restore(vocab_path)

  x_raw = df[selected[0]].apply(lambda x: clean_str(x)).tolist()
  x_test = [data_helper.clean_str(x) for x in x_raw]
  x_test = np.array(list(vocab_processor.transform(x_test)))
  logging.info('The number of x_test: {}'.format(len(x_test)))

  y_test = None
  y_raw = df[selected[1]].apply(lambda y: label_dict[y]).tolist()
  y_test = np.array(y_raw)
  logging.info('The number of y_test: {}'.format(len(y_test)))
  batches = data_helper.batch_iter(list(zip(x_test,y_test)), params['batch_size'], 1, shuffle=False)
  all_predictions = []
  start = time.time()
  for test_batch in batches:
    x_test_batch, y_test_batch = zip(*test_batch)

    x_test_batch = np.array(x_test_batch)
    x_test_batch = numpy_array_int32(x_test_batch)

    y_test_batch = np.array(y_test_batch)
    y_test_batch = numpy_array_float(y_test_batch)
    print(type(y_test_batch))

    request = predict_pb2.PredictRequest()
    request.model_spec.name = 'saved_model'
    request.model_spec.signature_name = 'predict_label'

    request.inputs['input_x'].CopyFrom(tf.contrib.util.make_tensor_proto(x_test_batch))
    request.inputs['input_y'].CopyFrom(tf.contrib.util.make_tensor_proto(y_test_batch))
    print("lalalala")
    request.inputs['dropout_keep_prob'].CopyFrom(tf.contrib.util.make_tensor_proto(1.0))
    print("hahahahaha")

    result = stub.Predict(request, 10.0)  # 10 secs timeout
    print(result.outputs['accuracy/accuracy'])
    batch_predictions = (result.outputs['output/predictions'].int64_val)
    all_predictions = np.concatenate([all_predictions, batch_predictions])
    print(batch_predictions)
    print(result)
    print("batch_predictions",batch_predictions)
    print("all_predictions",all_predictions )
  print(time.time() - start)

  if y_test is not None:
    y_test = np.argmax(y_test, axis=1)
    correct_predictions = sum(all_predictions == y_test)
    print(correct_predictions)

    for idx, example in enumerate(test_examples):
      print(idx, '\t', example)

    with open('./data/small_samples_prediction.json', 'w') as outfile:
      json.dump(test_examples.to_json(), outfile, indent=4)

    logging.critical('The accuracy is: {}'.format(correct_predictions / float(len(y_test))))
    logging.critical('The prediction is complete')

if __name__ == '__main__':
  tf.app.run()

**Comment by Aurora11111** [comment]
if you know any about the problem ,can you share some slotion with me?

**Comment by Aurora11111** [comment]
Amazing !I  work it out! I just add server code in train ,here the code below:
import os
import sys
import json
import time
import logging
import data_helper
import numpy as np
import tensorflow as tf
from text_cnn import TextCNN
from tensorflow.contrib import learn
from sklearn.model_selection import train_test_split

logging.getLogger().setLevel(logging.INFO)


def train_cnn():
    """Step 0: load sentences, labels, and training parameters"""
    train_file = "./data.csv"
    x_raw, y_raw, df, labels = data_helper.load_data_and_labels(train_file)
    print(x_raw)
    parameter_file = "./parameters.json"
    params = json.loads(open(parameter_file).read())

    """Step 1: pad each sentence to the same length and map each word to an id"""
    max_document_length = max([len(x.split(' ')) for x in x_raw])
    logging.info('The maximum length of all sentences: {}'.format(max_document_length))
    vocab_processor = learn.preprocessing.VocabularyProcessor(max_document_length)
    x = np.array(list(vocab_processor.fit_transform(x_raw)))
    y = np.array(y_raw)

    # print(x.shape)  (10146, 2976)
    print(type(x))
    # print(y.shape)  (10146, 8)
    print(type(y))
    """Step 2: split the original dataset into train and test sets"""
    x_, x_test, y_, y_test = train_test_split(x, y, test_size=0.1, random_state=42)

    """Step 3: shuffle the train set and split the train set into train and dev sets"""
    shuffle_indices = np.random.permutation(np.arange(len(y_)))
    x_shuffled = x_[shuffle_indices]
    y_shuffled = y_[shuffle_indices]
    x_train, x_dev, y_train, y_dev = train_test_split(x_shuffled, y_shuffled, test_size=0.1)

    """Step 4: save the labels into labels.json since predict.py needs it"""
    with open('./labels.json', 'w') as outfile:
        json.dump(labels, outfile, indent=4)

    logging.info('x_train: {}, x_dev: {}, x_test: {}'.format(len(x_train), len(x_dev), len(x_test)))
    logging.info('y_train: {}, y_dev: {}, y_test: {}'.format(len(y_train), len(y_dev), len(y_test)))

    """Step 5: build a graph and cnn object"""
    #print(x_train.shape[1]) 2976
    #print(y_train.shape[1]) 8
    graph = tf.Graph()
    with graph.as_default():
        session_conf = tf.ConfigProto(allow_soft_placement=True, log_device_placement=False)
        sess = tf.Session(config=session_conf)
        with sess.as_default():
            cnn = TextCNN(
                sequence_length=x_train.shape[1],
                num_classes=y_train.shape[1],
                vocab_size=len(vocab_processor.vocabulary_),
                embedding_size=params['embedding_dim'],
                filter_sizes=list(map(int, params['filter_sizes'].split(","))),
                num_filters=params['num_filters'],
                l2_reg_lambda=params['l2_reg_lambda'])

            global_step = tf.Variable(0, name="global_step", trainable=False)
            optimizer = tf.train.AdamOptimizer(1e-3)
            grads_and_vars = optimizer.compute_gradients(cnn.loss)
            train_op = optimizer.apply_gradients(grads_and_vars, global_step=global_step)

            timestamp = str(int(time.time()))
            out_dir = os.path.abspath(os.path.join(os.path.curdir, "trained_model_" + timestamp))

            checkpoint_dir = os.path.abspath(os.path.join(out_dir, "checkpoints"))
            checkpoint_prefix = os.path.join(checkpoint_dir, "model")
            if not os.path.exists(checkpoint_dir):
                os.makedirs(checkpoint_dir)
            variable_averages = tf.train.ExponentialMovingAverage(0.9997)

            variables_to_restore = variable_averages.variables_to_restore()
            saver = tf.train.Saver(variables_to_restore)
            #saver = tf.train.Saver()

            # One training step: train the model with one batch
            def train_step(x_batch, y_batch):
                feed_dict = {
                    cnn.input_x: x_batch,
                    cnn.input_y: y_batch,
                    cnn.dropout_keep_prob: params['dropout_keep_prob']
                            }
                _, step, loss, acc = sess.run([train_op, global_step, cnn.loss, cnn.accuracy], feed_dict)

            # One evaluation step: evaluate the model with one batch
            def dev_step(x_batch, y_batch):
                feed_dict = {cnn.input_x: x_batch, cnn.input_y: y_batch, cnn.dropout_keep_prob: 1.0}
                #feed_dict = {cnn.input_x: x_batch, cnn.input_y: y_batch}
                step, loss, acc, num_correct = sess.run([global_step, cnn.loss, cnn.accuracy, cnn.num_correct], feed_dict)
                return num_correct

            # Save the word_to_id map since predict.py needs it
            vocab_processor.save(os.path.join(out_dir, "vocab.pickle"))
            sess.run(tf.global_variables_initializer())

            # Training starts here
            train_batches = data_helper.batch_iter(list(zip(x_train, y_train)), params['batch_size'], params['num_epochs'])
            best_accuracy, best_at_step = 0, 0

            """Step 6: train the cnn model with x_train and y_train (batch by batch)"""
            for train_batch in train_batches:
                x_train_batch, y_train_batch = zip(*train_batch)
                train_step(x_train_batch, y_train_batch)
                current_step = tf.train.global_step(sess, global_step)

                """Step 6.1: evaluate the model with x_dev and y_dev (batch by batch)"""
                if current_step % params['evaluate_every'] == 0:
                    dev_batches = data_helper.batch_iter(list(zip(x_dev, y_dev)), params['batch_size'], 1)
                    total_dev_correct = 0
                    for dev_batch in dev_batches:
                        x_dev_batch, y_dev_batch = zip(*dev_batch)
                        num_dev_correct = dev_step(x_dev_batch, y_dev_batch)
                        total_dev_correct += num_dev_correct

                    dev_accuracy = float(total_dev_correct) / len(y_dev)
                    logging.critical('Accuracy on dev set: {}'.format(dev_accuracy))

                    """Step 6.2: save the model if it is the best based on accuracy on dev set"""
                    if dev_accuracy >= best_accuracy:
                        best_accuracy, best_at_step = dev_accuracy, current_step

                        path = saver.save(sess, checkpoint_prefix, global_step=current_step)
                        logging.critical('Saved model at {} at step {}'.format(path, best_at_step))
                        logging.critical('Best accuracy is {} at step {}'.format(best_accuracy, best_at_step))

            """Step 7: predict x_test (batch by batch)"""
            test_batches = data_helper.batch_iter(list(zip(x_test, y_test)), params['batch_size'], 1)
            total_test_correct = 0
            for test_batch in test_batches:
                x_test_batch, y_test_batch = zip(*test_batch)
                num_test_correct = dev_step(x_test_batch, y_test_batch)
                total_test_correct += num_test_correct

            test_accuracy = float(total_test_correct) / len(y_test)
            #logging.critical('Accuracy on test set is {} based on the best model {}'.format(test_accuracy, path))
            print(test_accuracy)
            logging.critical('The training is complete')




            import shutil


            export_path = os.path.join(
                tf.compat.as_bytes( 'model'),
                tf.compat.as_bytes(str(1)))
            if os.path.exists(export_path):
                shutil.rmtree(export_path)
            builder = tf.saved_model.builder.SavedModelBuilder(export_path)
            model_inputx = tf.saved_model.utils.build_tensor_info(cnn.input_x)
            model_dropout = tf.saved_model.utils.build_tensor_info(
                cnn.dropout_keep_prob)
            model_predictions = tf.saved_model.utils.build_tensor_info(cnn.predictions)

            prediction_signature = tf.saved_model.signature_def_utils.build_signature_def(
                inputs={'input_x': model_inputx,
                        #'input_y': model_inputy,

                        'dropout_keep_prob': model_dropout
                        },
                outputs={'output/predictions': model_predictions,
                         #'accuracy/accuracy': model_accuracy
                         },
                method_name=tf.saved_model.signature_constants.PREDICT_METHOD_NAME)

            # tensor_info_x = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('input_x:0'))
            # tensor_info_y = tf.saved_model.utils.build_tensor_info(graph.get_tensor_by_name('output/predictions:0'))

            # prediction_signature = (
            #     tf.saved_model.signature_def_utils.build_signature_def(
            #         inputs={'input_x': tensor_info_x},
            #         outputs={'output/predictions': tensor_info_y},
            #         method_name=tf.saved_model.signature_constants.PREDICT_METHOD_NAME))
            legacy_init_op = tf.group(tf.tables_initializer(),
                                      name='legacy_init_op')
            builder.add_meta_graph_and_variables(
                sess,
                [tf.saved_model.tag_constants.SERVING],
                signature_def_map={'predict_label': prediction_signature},
                legacy_init_op=legacy_init_op
            )
            # Save the model so we can serve it with a model server :)
            builder.save()


if __name__ == '__main__':
    # python3 train.py ./data/consumer_complaints.csv.zip ./parameters.json
    train_cnn()

**Comment by srihari-humbarwadi** [comment]
I am facing a similar issue when i try to freeze xception pretrained

**Comment by Aurora11111** [comment]
@srihari-humbarwadi 
I work it out!
It's the inputs and outputs in tf.saved_model.signature_def_utils.build_signature_def we use ,you must define the inputs and outputs correct, you can try as me in the comments3 below

**Comment by srihari-humbarwadi** [comment]
@Aurora11111 yeah I got it working too. i was missing some preprocessing on image
